# SQL

## Описание проекта 

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.

Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Ваша первая задача как аналитика — проанализировать базу данных.
В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

## Описание данных

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

## Задание

- Посчитайте, сколько книг вышло после 1 января 2000 года;
- Для каждой книги посчитайте количество обзоров и среднюю оценку;
- Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;
- Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
- Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

## Загрузка таблиц

In [1]:
# импортируем библиотеки
import pandas as pd
from pandas.io import sql 
from sqlalchemy import create_engine

In [2]:
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
 'pwd': 'Sdf4$2;d-d30pp', # пароль
 'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
 'port': 6432, # порт подключения
 'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

### Таблица books

In [3]:
# делаю запрос на таблицу books
query =    '''
            SELECT * FROM books
        '''
books = pd.io.sql.read_sql(query, con = engine)
books.head()

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [4]:
# смотрю инфо
books.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
book_id             1000 non-null int64
author_id           1000 non-null int64
title               1000 non-null object
num_pages           1000 non-null int64
publication_date    1000 non-null object
publisher_id        1000 non-null int64
dtypes: int64(4), object(2)
memory usage: 47.0+ KB


### Таблица authors

In [5]:
# делаю запрос на таблицу authors
query =    '''
            SELECT * FROM authors
        '''
authors = pd.io.sql.read_sql(query, con = engine)
authors.head()

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [6]:
# смотрю инфо
authors.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
author_id    636 non-null int64
author       636 non-null object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB


### Таблица publishers

In [7]:
# делаю запрос на таблицу publishers
query =    '''
            SELECT * FROM publishers
        '''
publishers = pd.io.sql.read_sql(query, con = engine)
publishers.head()

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [8]:
# смотрю инфо
publishers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
publisher_id    340 non-null int64
publisher       340 non-null object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB


### Таблица ratings

In [9]:
# делаю запрос на таблицу ratings
query =    '''
            SELECT * FROM ratings
        '''
ratings = pd.io.sql.read_sql(query, con = engine)
ratings.head()

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [10]:
ratings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
rating_id    6456 non-null int64
book_id      6456 non-null int64
username     6456 non-null object
rating       6456 non-null int64
dtypes: int64(3), object(1)
memory usage: 201.9+ KB


### Таблица reviews

In [11]:
# делаю запрос на таблицу reviews
query =    '''
SELECT * 
FROM reviews
        '''
reviews = pd.io.sql.read_sql(query, con = engine)
reviews.head()

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

In [12]:
reviews.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
review_id    2793 non-null int64
book_id      2793 non-null int64
username     2793 non-null object
text         2793 non-null object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB


## Ответы на вопросы

### Сколько книг вышло после 1 января 2000 ?

In [14]:
# делаю запрос
count_books  = '''
SELECT 
  COUNT(book_id) as cnt
FROM 
  books
WHERE 
  publication_date > '2000-01-01'
'''

In [15]:
count_books = pd.read_sql(count_books, con = engine)
count_books

cnt
0  819

После 1 января 2000 вышло 819 книг.

### Какое количество обзоров и какая средняя оценка?

In [16]:
# делаю запрос
count_reviews =  '''
SELECT
  books.title,
  COUNT(DISTINCT reviews.review_id) AS cnt_reviews,  
  AVG(ratings.rating) AS avg_ratings      
FROM 
  books
  LEFT JOIN ratings ON ratings.book_id = books.book_id
  LEFT JOIN reviews ON reviews.book_id= books.book_id
GROUP BY
  books.book_id
ORDER BY
  cnt_reviews DESC
    
'''

In [17]:
count_reviews = pd.io.sql.read_sql(count_reviews, con = engine)
count_reviews.head(10)

title  cnt_reviews  avg_ratings
0                            Twilight (Twilight  #1)            7     3.662500
1                                Water for Elephants            6     3.977273
2                                   The Glass Castle            6     4.206897
3  Harry Potter and the Prisoner of Azkaban (Harr...            6     4.414634
4  The Curious Incident of the Dog in the Night-Time            6     4.081081
5             The Da Vinci Code (Robert Langdon  #2)            6     3.830508
6                                      The Alchemist            6     3.789474
7                The Hobbit  or There and Back Again            6     4.125000
8                                     The Book Thief            6     4.264151
9  The Lightning Thief (Percy Jackson and the Oly...            6     4.080645

## Какое издетельство выпустило наибольшее число книг толще 50 страниц?

In [19]:
# делаю запрос
pages = '''
SELECT
  publishers.publisher,
  COUNT(books.book_id) AS cnt
FROM    
  books
  LEFT JOIN publishers ON books.publisher_id = publishers.publisher_id
WHERE 
  num_pages > 50
GROUP BY
  publishers.publisher
ORDER BY
  cnt DESC
'''

In [20]:
pub_cnt = pd.io.sql.read_sql(pages, con = engine)
pub_cnt

publisher  cnt
0                                        Penguin Books   42
1                                              Vintage   31
2                             Grand Central Publishing   25
3                                     Penguin Classics   24
4                                     Ballantine Books   19
..                                                 ...  ...
329                                         Turtleback    1
330  Atheneum Books for Young Readers: Richard Jack...    1
331                                     Penguin Signet    1
332                                    Victor Gollancz    1
333                      Harvard Business Review Press    1

[334 rows x 2 columns]

Наибольшое количество выпущенных книг больше 50 страниц у издательства Penguin Books - 42 

### У какого автора наиболее высокая оценка среди тех у кого их больше 50?

In [21]:
author  = '''
SELECT
  authors.author,
  AVG(ratings.rating) AS avg_rating
FROM 
  books
  LEFT JOIN authors ON books.author_id = authors.author_id
  LEFT JOIN ratings ON ratings.book_id = books.book_id
WHERE
  books.book_id IN
    (SELECT
      books.book_id
     FROM
      books    
      LEFT JOIN ratings ON ratings.book_id = books.book_id     
     GROUP BY
      books.book_id
     HAVING 
      COUNT(ratings.rating_id) > 50)
GROUP BY
  authors.author  
ORDER BY 
  avg_rating DESC 
'''

In [22]:
ratings_cnt = pd.io.sql.read_sql(author, con = engine)
ratings_cnt

author  avg_rating
0                          J.K. Rowling/Mary GrandPré    4.287097
1                   Markus Zusak/Cao Xuân Việt Khương    4.264151
2                                      J.R.R. Tolkien    4.246914
3                                   Louisa May Alcott    4.192308
4                                        Rick Riordan    4.080645
5                                     William Golding    3.901408
6                                       J.D. Salinger    3.825581
7            Paulo Coelho/Alan R. Clarke/Özdemir İnce    3.789474
8   William Shakespeare/Paul Werstine/Barbara A. M...    3.787879
9                                          Lois Lowry    3.750000
10                                          Dan Brown    3.741259
11            George Orwell/Boris Grabnar/Peter Škerl    3.729730
12                                    Stephenie Meyer    3.662500
13                                     John Steinbeck    3.622951

Наиболее высоко оцениваемой стала J.K. Rowling c иллюстраторкой её книг Mary GrandPré.

### Какое среднее количество обзоров от пользователей, которые поставили больше 50 оценок?

In [23]:
avg_review  = '''
SELECT
  SUM(subq.cnt) / COUNT(subq.username) AS avg_cnt_reviews
FROM
  (SELECT
    reviews.username,
    COUNT(reviews.review_id) AS cnt
   FROM
    reviews
   WHERE reviews.username IN
    (SELECT
      ratings.username
     FROM
      ratings
      LEFT JOIN books ON books.book_id = ratings.book_id
     GROUP BY
        ratings.username
     HAVING
        COUNT(ratings.rating_id) > 50)
     GROUP BY
        reviews.username) AS subq

       '''

In [24]:
avg_review  = pd.read_sql(avg_review, con = engine)
avg_review 

avg_cnt_reviews
0        24.333333

Среднее количество обзоров от пользователей, которые поставили более 50 оценок - 24.

## Ответы:

* После 1 января 2000 вышло 819 книг.
* Самое большое количество обзоров у Twilight (Twilight #1) - 7.
* Наибольшое количество выпущенных книг больше 50 страниц у издательства Penguin Books - 42.
* Наиболее высоко оцениваемой стала J.K. Rowling c иллюстраторкой её книг Mary GrandPré.
* Среднее количество обзоров от пользователей, которые поставили более 50 оценок - 24.